app/index.js

```diff
@@ -267,20 +267,50 @@ server.route({
       let { query } = request;
       let targetYear = null;
 
-      // Check for redis data with this query.
+      // Generate cache key from query params
       const cacheKey = JSON.stringify({ id, query });
-      if (redisEnabled) {
-        const cachedData = await redis.get(cacheKey);
+      if (cache.enabled) {
+        // Load cache entry
+        let cachedData = await redisClient.getObject(cacheKey);
+
+        // If cache entry exists
         if (cachedData) {
-          // Once the data is requested, store for a week
-          await redis.expire(cacheKey, redisCacheTtl);
+          // Check every second if query is running. As database queries can
+          // take a while, this avoids starting the same query in parallel.
+          let duration = 0;
+          let step = 1000;
+          while (cachedData.runningDbQuery) {
+            // If query is running for too long, abort cache load and throw
+            // error.
+            if (duration > cache.loadTimeout) {
```

Comment:
This load timeout should be < 30 seconds, otherwise we're just going to be waiting on a result that's just going to get lost in the ELB. Which neatly solves the "Return a Gateway timeout error" but may not be what we want.